packages/apps/dashboard/ui-2024/src/features/Leaderboard/hooks/useDatagridOptions.tsx

```diff
@@ -0,0 +1,128 @@
+import { Typography } from '@mui/material';
+import {
+  GridColDef,
+  GridRenderCellParams,
+  GridSortModel,
+  GridCallbackDetails,
+  GridValidRowModel,
+} from '@mui/x-data-grid';
+import { GridApiCommunity } from '@mui/x-data-grid/models/api/gridApiCommunity';
+import { useMemo, useCallback, MutableRefObject } from 'react';
+import { useBreakPoints } from '@utils/hooks/use-is-mobile';
+
+interface IColPossibleOptions {
+  isRowIdx?: boolean;
+  pinnedColumnName?: string;
+}
+
+interface IProps<T extends GridValidRowModel> extends IColPossibleOptions {
+  columns: GridColDef<T>[];
+  apiRef: MutableRefObject<GridApiCommunity>;
+}
+
+export function useDatagridOptions<T extends GridValidRowModel>({
+  apiRef,
+  columns,
+  isRowIdx,
+  pinnedColumnName,
+}: IProps<T>) {
+  const {
+    mobile: { isMobile },
+  } = useBreakPoints();
+
+  const handlePinnedColOption = useCallback(
+    (cols: GridColDef<T>[]) =>
+      cols.map((col) =>
+        col.field === pinnedColumnName
+          ? {
+              ...col,
+              headerClassName: isMobile
+                ? 'home-page-table-header pinned-column--header'
+                : 'home-page-table-header',
+              cellClassName: isMobile ? 'pinned-column--cell' : '',
+            }
+          : col
+      ),
+    [isMobile, pinnedColumnName]
+  );
+
+  const handleRowOrderNum = useCallback(
+    (cols: GridColDef<T>[]) => {
+      return isMobile
+        ? cols
+        : [
+            {
+              field: 'id',
+              headerName: '',
+              width: 30,
+              headerClassName: 'home-page-table-header',
+              sortable: false,
+              renderCell: (params: GridRenderCellParams) => (
+                <Typography
+                  variant="body1"
+                  height="100%"
+                  display="flex"
+                  alignItems="center"
+                >
+                  {params.api.state.sorting.sortedRows.indexOf(params.id) + 1}
+                </Typography>
+              ),
+            },
+            ...cols,
+          ];
+    },
+    [isMobile]
+  );
+
+  const handleColumnOptions = useCallback(
+    (cols: GridColDef<T>[]) => {
+      let colsWithOptions = cols;
+      if (isRowIdx) {
+        colsWithOptions = handleRowOrderNum(colsWithOptions);
+      }
+      if (pinnedColumnName) {
+        colsWithOptions = handlePinnedColOption(colsWithOptions);
+      }
+
+      return colsWithOptions;
+    },
+    [handlePinnedColOption, handleRowOrderNum, isRowIdx, pinnedColumnName]
+  );
+
+  const handleSortModelChange = useCallback(
+    (_model: GridSortModel, details: GridCallbackDetails) => {
+      details.api.getSortedRows().forEach((row, idx) => {
+        apiRef.current.updateRows([{ ...row, id: idx }]);
```

Comment:
That's definitely better and cleaner than implicit re-render! Love it!

I assume that calling `updateRows` for each row will trigger re-render multiple times, so maybe we
```
const sortedRows = details.api.getSortedRows();
apiRef.current.updateRows(sortedRows.map((row, idx) => ({ ...row, id: idx })))
```
wdyt?

Reply:
Yes that would be definitely better approach but unfortunately we are limited here because of a plan.
![Zrzut ekranu 2024-11-21 o 09 37 33](https://github.com/user-attachments/assets/965c7ecb-59d2-4559-aabe-aeeec123decb)

But still I improved it a bit thanks to your comment, there is no need to update id

Reply:
https://mui.com/x/react-data-grid/row-updates/#the-updaterows-method

Reply:
What a tricky surprise from them 😒